### ISSUE TITLE
Enable target-required-features with --all-features

### ISSUE THREAD

**Issue by pinkforest** [rust-lang/cargo#10332]
### Problem

If you have multiple [[targets]]

It can be pretty exhausting to manually juggle with the individual deviating --features sets at CLI under [[target]]

Especially as --all-features does not take into account / resolve the dependency features which in turn overrides explicit --features [another issue]

**There was a previous issue and stale PR to trigger dependency features under [[target]]**
https://github.com/rust-lang/cargo/issues/1570
https://github.com/rust-lang/cargo/pull/2325

**But required-features was decided instead**
https://github.com/rust-lang/cargo/pull/3667

And without separating things into workspaces relying on `--features` can become big hurdlesome list.

### Proposed Solution

**Option 0 - Do nothing or Status Quo**

Yes I could make for example feature-relay hack say under the package
```toml
[features]
example-features = ["dependency/*"]
```
This would enable them via --all-features but it's not elegant way as the user who runs these examples would just do cargo run --example foo without too much thinking.

**Option 1 - --all-features to enable contextual required-features**

If `cargo run --all-features --example foo` is ran - 

I would expect cargo to enable the required-features on dependencies contextually on per-target basis.

**Option 2 - features under [[target]]**

Re-think #1570 and perhaps support feature enable trigger under [[target]]

**Option 3 - fix --features with --all-features**

See #10333 

### Notes

This was discussed in zulip 
https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/.60--all-features.60.20ignores.20.60--features.60.20and.20suggests.20using.20it/near/269409108

**Comment by ehuss** [comment]
Sorry, I'm having a hard time trying to understand what this issue is asking for.  Can you show an example project, and what is difficult to manage in it? Which cargo commands are you trying to run?  I'm also not quite following what is meant by `[[target]]`, is that referring to tables like '[[example]]`?

**Comment by pinkforest** [comment]
@ehuss I worked this in collaboration with Issue #10333 and is heavily related to that issue which fixed that you can enable additional (dependency) features along with --all-features that represent only the root package features.

Also we discussed this in zulip  which led to these two issues -
https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/.60--all-features.60.20ignores.20.60--features.60.20and.20suggests.20using.20it/near/269409108

My additional take on to that was since every [[target]] has required features under -

One is supposed to use those required-features to signal what additional features the [[target]] requires -

E.g. an example that requires different featureset dependencies -

We can't just say --all-features and be good since you have to manually toggle every feature on via --feature that the [[target]] asks via required features under it which seems counterintuitive.

Why can't we just just say --all-features and then the dependency resolution figures out the additional required features to toggle on via required-features under [[targets]] that will be included in our run?

If you don't manually enable --feature x where the x was required it will throw an error that the feature was not enabled despite --all-features on.

**Comment by ehuss** [comment]
I'm still having a hard time trying to follow, but I believe `--all-features` will not be able to enable features on all dependencies.  

If you're asking that you want building an example to automatically enable a feature, there are other existing issues like #1982 (and accompanying RFCs like https://github.com/rust-lang/rfcs/pull/3020 or https://github.com/rust-lang/rfcs/pull/2887).  That would make it so that in your example, `cargo build --examples` would provide a way to force enable features (instead of the current behavior of `required-features` which just makes the example skip if not enabled).

Does that cover your use case?

**Comment by pinkforest** [comment]
So looks like @pksunkara was going to work on changing something - 

It is unclear what was actually really proposed or decided if any for the future of this considering the history of how required-features came to be:

https://github.com/rust-lang/cargo/issues/1570
https://github.com/rust-lang/cargo/pull/2325
https://github.com/rust-lang/cargo/pull/3667

I was merely proposing to work with required-features under [[target]] without changing it considering the above and without changing how the resolver works except adding the features from requires-features automatically instead of raising errors.

e.g. if I want to build all [[bin]] targets I shouldn't be needing to enable the required features manually - the error seems redundant that complains some required-feature is not enabled?

Bottom line:

So is cargo going to preserve required-features or is it going to be changed? RFCs don't make it clear neither from porposal or decision PoV.
